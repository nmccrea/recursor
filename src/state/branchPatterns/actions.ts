import { createAction } from "@reduxjs/toolkit"
import {
  NewBranchPattern,
  BranchPatternId,
  Translation,
  Scale,
  Angle,
  Depth,
} from "./types"

/**
 * Creates an action containing a new branch pattern to be added to the list of branch patterns.
 *
 * @param branchPattern - The new branch pattern.
 *
 * @returns A payload action whose payload is the new branch pattern to add.
 */
const addOne = createAction<NewBranchPattern>("branchPattern/ADD_ONE")

/**
 * Creates an action identifying an existing branch pattern to be destroyed.
 *
 * @param branchPatternId - The ID of the branch pattern to remove.
 *
 * @returns A payload action whose payload conforms to `@reduxjs/toolkit`'s `entityAdapter.removeOne()` API.
 */
const removeOne = createAction<BranchPatternId>("branchPattern/REMOVE_ONE")

/**
 * Creates an action representing a change to the identified branch pattern's translation.
 *
 * @param id - The ID of the branch pattern to update.
 * @param translation - The new translation value.
 *
 * @returns A payload action whose payload conforms to `@reduxjs/toolkit`'s `entityAdapter.updateOne()` API.
 */
const setTranslation = createAction(
  "branchPattern/SET_TRANSLATION",
  (id: BranchPatternId, translation: Translation) => ({
    payload: {
      id,
      changes: { translation },
    },
  })
)

/**
 * Creates an action representing a change to the identified branch pattern's scale.
 *
 * @param id - The ID of the branch pattern to update.
 * @param scale - The new scale value.
 *
 * @returns A payload action whose payload conforms to `@reduxjs/toolkit`'s `entityAdapter.updateOne()` API.
 */
const setScale = createAction(
  "branchPattern/SET_SCALE",
  (id: BranchPatternId, scale: Scale) => ({
    payload: {
      id,
      changes: { scale },
    },
  })
)

/**
 * Creates an action representing a change to the identified branch pattern's angle.
 *
 * @param id - The ID of the branch pattern to update.
 * @param angle - The new angle value.
 *
 * @returns A payload action whose payload conforms to `@reduxjs/toolkit`'s `entityAdapter.updateOne()` API.
 */
const setAngle = createAction(
  "branchPattern/SET_ANGLE",
  (id: BranchPatternId, angle: Angle) => ({
    payload: {
      id,
      changes: { angle },
    },
  })
)

/**
 * Creates an action representing a change to the identified branch pattern's depth.
 *
 * @param id - The ID of the branch pattern to update.
 * @param depth - The new depth value.
 *
 * @returns A payload action whose payload conforms to `@reduxjs/toolkit`'s `entityAdapter.updateOne()` API.
 */
const setDepth = createAction(
  "branchPattern/SET_DEPTH",
  (id: BranchPatternId, depth: Depth) => ({
    payload: {
      id,
      changes: { depth },
    },
  })
)

export { addOne, removeOne, setTranslation, setScale, setAngle, setDepth }
