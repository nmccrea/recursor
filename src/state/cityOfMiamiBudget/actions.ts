/**
 * TODO: type
 * TODO: test
 * TODO: doc
 */

import { createAction } from "@reduxjs/toolkit"

const fetchDataStart = createAction("FETCH_DATA/start")

const fetchDataSuccess = createAction<string>("FETCH_DATA/success")

export { fetchDataStart, fetchDataSuccess }
