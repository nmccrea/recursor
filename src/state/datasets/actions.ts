/**
 * TODO: type
 * TODO: test
 * TODO: doc
 */

import { createAction } from "@reduxjs/toolkit"
import { DatasetId, FetchResult } from "./types"

const fetchDatasetStart = createAction<DatasetId>(
  "datasets/FETCH_DATASET/start"
)

const fetchDatasetSuccess = createAction<FetchResult>(
  "datasets/FETCH_DATASET/success"
)

export { fetchDatasetStart, fetchDatasetSuccess }
