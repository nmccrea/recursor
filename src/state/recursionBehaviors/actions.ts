import { createAction } from "@reduxjs/toolkit"
import { RecursionBehaviorId, Translation, Scale, Angle, Depth } from "./types"

/**
 * Creates an action representing a change to the identified recursion behavior's translation.
 *
 * @param id - The ID of the recursion behavior to update.
 * @param translation - The new translation value.
 *
 * @returns A payload action whose payload conforms to `@reduxjs/toolkit`'s `entityAdapter.updateOne()` API.
 */
const setTranslation = createAction(
  "treeFractal/SET_TRANSLATION",
  (id: RecursionBehaviorId, translation: Translation) => ({
    payload: {
      id,
      changes: { translation },
    },
  })
)

/**
 * Creates an action representing a change to the identified recursion behavior's scale.
 *
 * @param id - The ID of the recursion behavior to update.
 * @param scale - The new scale value.
 *
 * @returns A payload action whose payload conforms to `@reduxjs/toolkit`'s `entityAdapter.updateOne()` API.
 */
const setScale = createAction(
  "treeFractal/SET_SCALE",
  (id: RecursionBehaviorId, scale: Scale) => ({
    payload: {
      id,
      changes: { scale },
    },
  })
)

/**
 * Creates an action representing a change to the identified recursion behavior's angle.
 *
 * @param id - The ID of the recursion behavior to update.
 * @param angle - The new angle value.
 *
 * @returns A payload action whose payload conforms to `@reduxjs/toolkit`'s `entityAdapter.updateOne()` API.
 */
const setAngle = createAction(
  "treeFractal/SET_ANGLE",
  (id: RecursionBehaviorId, angle: Angle) => ({
    payload: {
      id,
      changes: { angle },
    },
  })
)

/**
 * Creates an action representing a change to the identified recursion behavior's depth.
 *
 * @param id - The ID of the recursion behavior to update.
 * @param depth - The new depth value.
 *
 * @returns A payload action whose payload conforms to `@reduxjs/toolkit`'s `entityAdapter.updateOne()` API.
 */
const setDepth = createAction(
  "treeFractal/SET_DEPTH",
  (id: RecursionBehaviorId, depth: Depth) => ({
    payload: {
      id,
      changes: { depth },
    },
  })
)

export { setTranslation, setScale, setAngle, setDepth }
