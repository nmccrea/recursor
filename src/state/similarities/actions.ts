import { createAction, PayloadAction } from "@reduxjs/toolkit"
import {
  SimilarityId,
  SimilarityInputs,
  SimilarityInputKey,
  Translation,
  Scale,
  Angle,
  Depth,
} from "./types"

/**
 * A payload action representing a change to a similarity input.
 */
type SimilarityInputAction<Key extends SimilarityInputKey> = PayloadAction<{
  id: SimilarityId
  changes: Pick<SimilarityInputs, Key>
}>

/**
 * An action creator for updating a similarity input. The function will already be bound to a particular similarity ID so only the new value is required.
 *
 * @param value - The new value of the input.
 *
 * @returns A similarity input action with the given value.
 */
export type SimilarityInputActionCreator<Key extends SimilarityInputKey> = (
  value: SimilarityInputs[Key]
) => SimilarityInputAction<Key>

/**
 * Creates an action representing the creation of a new similarity to be added to the list of similarities.
 *
 * @returns A payload action with an empty payload.
 */
const createOne = createAction<undefined>("similarities/CREATE_ONE")

/**
 * Creates an action identifying an existing similarity to be destroyed.
 *
 * @param similarityId - The ID of the similarity to remove.
 *
 * @returns A payload action containing the given ID.
 */
const removeOne = createAction<SimilarityId>("similarities/REMOVE_ONE")

/**
 * Creates an action representing a change to the identified similarity's translation.
 *
 * @param id - The ID of the similarity to update.
 * @param translation - The new translation value.
 *
 * @returns A similarity input action representing this change.
 */
const setTranslation = createAction(
  "similarities/SET_TRANSLATION",
  (id: SimilarityId, translation: Translation) => ({
    payload: {
      id,
      changes: { translation },
    },
  })
)

/**
 * Gets an action creator for setting the translation of the identified similarity.
 *
 * @param id - The ID of the similarity for which to create a translation setter.
 *
 * @returns A similarity input action creator.
 */
const getTranslationSetterFor = (
  id: SimilarityId
): SimilarityInputActionCreator<"translation"> => (translation) =>
  setTranslation(id, translation)

/**
 * Creates an action representing a change to the identified similarity's scale.
 *
 * @param id - The ID of the similarity to update.
 * @param scale - The new scale value.
 *
 * @returns A similarity input update representing this change.
 */
const setScale = createAction(
  "similarities/SET_SCALE",
  (id: SimilarityId, scale: Scale) => ({
    payload: {
      id,
      changes: { scale },
    },
  })
)

/**
 * Gets an action creator for setting the scale of the identified similarity.
 *
 * @param id - The ID of the similarity for which to create a scale setter.
 *
 * @returns A similarity input action creator.
 */
const getScaleSetterFor = (
  id: SimilarityId
): SimilarityInputActionCreator<"scale"> => (scale) => setScale(id, scale)

/**
 * Creates an action representing a change to the identified similarity's angle.
 *
 * @param id - The ID of the similarity to update.
 * @param angle - The new angle value.
 *
 * @returns A similarity input update representing this change.
 */
const setAngle = createAction(
  "similarities/SET_ANGLE",
  (id: SimilarityId, angle: Angle) => ({
    payload: {
      id,
      changes: { angle },
    },
  })
)

/**
 * Gets an action creator for setting the angle of the identified similarity.
 *
 * @param id - The ID of the similarity for which to create a angle setter.
 *
 * @returns A similarity input action creator.
 */
const getAngleSetterFor = (
  id: SimilarityId
): SimilarityInputActionCreator<"angle"> => (angle) => setAngle(id, angle)

/**
 * Creates an action representing a change to the identified similarity's depth.
 *
 * @param id - The ID of the similarity to update.
 * @param depth - The new depth value.
 *
 * @returns A similarity input update representing this change.
 */
const setDepth = createAction(
  "similarities/SET_DEPTH",
  (id: SimilarityId, depth: Depth) => ({
    payload: {
      id,
      changes: { depth },
    },
  })
)

/**
 * Gets an action creator for setting the depth of the identified similarity.
 *
 * @param id - The ID of the similarity for which to create a depth setter.
 *
 * @returns A similarity input action creator.
 */
const getDepthSetterFor = (
  id: SimilarityId
): SimilarityInputActionCreator<"depth"> => (depth) => setDepth(id, depth)

export type CreateOneAction = ReturnType<typeof createOne>
export type RemoveOneAction = ReturnType<typeof removeOne>
export type SetTranslationAction = ReturnType<typeof setTranslation>
export type SetScaleAction = ReturnType<typeof setScale>
export type SetAngleAction = ReturnType<typeof setAngle>
export type SetDepthAction = ReturnType<typeof setDepth>

export {
  createOne,
  removeOne,
  setTranslation,
  getTranslationSetterFor,
  setScale,
  getScaleSetterFor,
  setAngle,
  getAngleSetterFor,
  setDepth,
  getDepthSetterFor,
}
