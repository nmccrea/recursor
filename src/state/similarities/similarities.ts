import { createEntityAdapter } from "@reduxjs/toolkit"
import { Similarity } from "./types"

const entityAdapter = createEntityAdapter<Similarity>()
export default entityAdapter
export type Similarities = ReturnType<typeof entityAdapter.getInitialState>
