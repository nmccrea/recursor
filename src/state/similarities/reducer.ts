import { createReducer, nanoid, PayloadAction } from "@reduxjs/toolkit"
import similarities, { Similarities } from "./similarities"
import {
  addOne,
  removeOne,
  setTranslation,
  setScale,
  setAngle,
  setDepth,
} from "./actions"
import { SimilarityInputs } from "./types"

/**
 * A case reducer which creates a similarity out of the given similarity inputs and adds it to the list.
 */
const addOneWithUniqueId = (
  previousState: Similarities,
  action: PayloadAction<SimilarityInputs>
): Similarities => {
  const similarity = action.payload
  return similarities.addOne(previousState, {
    ...similarity,
    id: nanoid(),
  })
}

const reducer = createReducer(similarities.getInitialState(), (builder) =>
  builder
    .addCase(addOne, addOneWithUniqueId)
    .addCase(removeOne, similarities.removeOne)
    .addCase(setTranslation, similarities.updateOne)
    .addCase(setScale, similarities.updateOne)
    .addCase(setAngle, similarities.updateOne)
    .addCase(setDepth, similarities.updateOne)
)

export default reducer
