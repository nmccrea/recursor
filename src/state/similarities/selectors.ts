import { RootState } from "../store"
import similarities from "./similarities"
import { Selector } from "../store"
import { SimilarityId, SimilarityInputs, SimilarityInputKey } from "./types"

/**
 * A selector for getting a similarity input value. The function will already be bound to a particular similarity ID.
 */
export type SimilarityInputSelector<Key extends SimilarityInputKey> = Selector<
  SimilarityInputs[Key]
>

/**
 * Selects the similarities slice of the given state tree.
 */
const selectSimilarities = (state: RootState) => state.similarities

/**
 * Selects a list containing all of the existing similarities.
 * Results are ordered according to the entity adaptor's `ids` field.
 */
const selectAll = similarities.getSelectors(selectSimilarities).selectAll

/**
 * Selects the ordered list of similarity IDs.
 */
const selectIds = similarities.getSelectors(selectSimilarities).selectIds

/**
 * Returns a selector for the identified similarity.
 *
 * @param id - The ID of the similarity to select.
 *
 * @returns A selector which selects the identified similarity when called.
 */
const getSimilaritySelectorFor = (id: SimilarityId) => (state: RootState) =>
  similarities.getSelectors(selectSimilarities).selectById(state, id)

/**
 * Returns a selector for the identified similarity's translation.
 *
 * @param id - The ID of the similarity to select from.
 *
 * @returns A selector which returns the translation from the identified similarity when called.
 */
const getTranslationSelectorFor = (
  id: SimilarityId
): SimilarityInputSelector<"translation"> => (state) => {
  const similarity = similarities
    .getSelectors(selectSimilarities)
    .selectById(state, id)
  return similarity?.translation
}

/**
 * Returns a selector for the identified similarity's scale.
 *
 * @param id - The ID of the similarity to select from.
 *
 * @returns A selector which returns the scale from the identified similarity when called.
 */
const getScaleSelectorFor = (
  id: SimilarityId
): SimilarityInputSelector<"scale"> => (state) => {
  const similarity = similarities
    .getSelectors(selectSimilarities)
    .selectById(state, id)
  return similarity?.scale
}

/**
 * Returns a selector for the identified similarity's angle.
 *
 * @param id - The ID of the similarity to select from.
 *
 * @returns A selector which returns the angle from the identified similarity when called.
 */
const getAngleSelectorFor = (
  id: SimilarityId
): SimilarityInputSelector<"angle"> => (state) => {
  const similarity = similarities
    .getSelectors(selectSimilarities)
    .selectById(state, id)
  return similarity?.angle
}

/**
 * Returns a selector for the identified similarity's depth.
 *
 * @param id - The ID of the similarity to select from.
 *
 * @returns A selector which returns the depth from the identified similarity when called.
 */
const getDepthSelectorFor = (
  id: SimilarityId
): SimilarityInputSelector<"depth"> => (state) => {
  const similarity = similarities
    .getSelectors(selectSimilarities)
    .selectById(state, id)
  return similarity?.depth
}

export {
  selectAll,
  selectIds,
  getSimilaritySelectorFor,
  getTranslationSelectorFor,
  getScaleSelectorFor,
  getAngleSelectorFor,
  getDepthSelectorFor,
}
