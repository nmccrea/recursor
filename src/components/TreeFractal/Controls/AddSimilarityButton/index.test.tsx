import React from "react"
import { render, fireEvent, getByText } from "@testing-library/react"
import * as reactRedux from "react-redux"
import { AddOneAction } from "../../../../state/similarities/actions"
import AddsimilarityButton from "."

describe("`<AddSimilarityButton />`", () => {
  it("renders correctly", () => {
    jest.spyOn(reactRedux, "useDispatch").mockReturnValue(jest.fn())

    const { container } = render(<AddsimilarityButton />)

    expect(container).toMatchSnapshot()
  })

  it("responds correctly to a click event", () => {
    const mockDispatch = jest.fn()
    jest.spyOn(reactRedux, "useDispatch").mockReturnValue(mockDispatch)
    const { container } = render(<AddsimilarityButton />)

    fireEvent.click(getByText(container, "Add Similarity"))

    expect(mockDispatch).toHaveBeenCalledWith<[AddOneAction]>({
      type: "similarities/ADD_ONE",
      payload: {
        translation: 1,
        scale: 0.9,
        angle: 1.23425,
        depth: 4,
        color: "black",
      },
    })
  })
})
