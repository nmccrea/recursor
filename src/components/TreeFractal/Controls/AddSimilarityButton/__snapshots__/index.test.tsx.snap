// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`<AddSimilarityButton />\` renders correctly 1`] = `
<div>
  <button>
    Add Similarity
  </button>
</div>
`;
